### PR TITLE
Accept a num_elements_provider in FieldArray to specify a size in elements instead of bytes.

### LINKED ISSUE #34: Accept a "number of elements" argument for FieldArray
It's not uncommon to know how many elements are in an array without knowing the total size of the array in bytes. For instance, with an array of strings: Each string might know its own size, but the array only know how many strings exists.

Right now, it doesn't seem possible to model this kind of data structure.

### PR DESCRIPTION
**Pull request by rcfox**
After some thought, I was able to come up with a way to resolve #34 myself.

### CODE REVIEW

**Comment by rcfox** [comment]
Not sure what happened with the coverage build, but I don't think it's my fault.

    coverage runtests: commands[1] | coveralls
    Traceback (most recent call last):
      File ".tox/coverage/bin/coveralls", line 11, in <module>
        sys.exit(wear())
      File "/home/travis/build/digidotcom/python-suitcase/.tox/coverage/lib/python3.5/site-packages/coveralls/__init__.py", line 79, in wear
        from coveralls.control import coveralls
      File "/home/travis/build/digidotcom/python-suitcase/.tox/coverage/lib/python3.5/site-packages/coveralls/control.py", line 1, in <module>
        from coverage.control import Coverage
    ImportError: cannot import name 'Coverage'
    ERROR: InvocationError: '/home/travis/build/digidotcom/python-suitcase/.tox/coverage/bin/coveralls'

File: suitcase/fields.py [hunk 1]
```diff
@@ -897,6 +912,14 @@ def unpack(self, data, **kwargs):
             self._value.append(structure)
             if data == b"":
                 break
+            if len(self._value) == self.num_elements:
+                break
+
+        if self.num_elements is not None and len(self._value) != self.num_elements:
+            raise SuitcaseParseError("Expected %r elements but received %r." %
```
Comment (mikewadsten) [inline suitcase/fields.py]: I'd recommend using %s and %d, respectively, in this format string. Using %r will add single quotes which could be confusing.

**Comment by posborne** [comment]
I actually prefer `%r` in this case as `%s` formatting will not show some characters (binary non-printable).  In addition, things like newlines will just be included and output when printed which can be very confusing.

**Comment by posborne** [comment]
@rcfox I can look at the coverage stuff and will look at this PR in the next few days (hopefully) once I get back from traveling.

**Comment by mikewadsten** [comment]
@posborne You're absolutely right on the `%r` stuff. I had it in my head that `"%r %r" % (None, 1)` yields the string `'None' '1'`, when that's obviously not the case.

**Comment by posborne** [comment]
Build failure should be fixed by https://github.com/digidotcom/python-suitcase/pull/41 and I think this is a pretty valuable change to have in.  I'm going to go ahead and include it.

Thanks and sorry for the long delay!

**Pull request merged**